docs/adr/0002-project-locks.md

```diff
@@ -0,0 +1,169 @@
+- add non-default workspaces to workflow hooks [#2882](https://github.com/runatlantis/atlantis/pull/2882)
+    - reintroduces #2180 changes to working_dir
+    - run hooks on the default workspace
+    - also run hooks on every project found
+    - significant increase in execution time, especially for Terragrunt users relying on terragrunt-atlantis-config
+- reduce the amount of git clone operations [#2921](https://github.com/runatlantis/atlantis/pull/2921)
+    - started to try to reduce execution time on #2882 implementation
+    - attempts to utilize [TF_DATA_DIR](https://developer.hashicorp.com/terraform/cli/config/environment-variables#tf_data_dir) for workspaces and remove the workspace for the lock. 
+    - Clones only at BaseRepo/PR# and moves workspaces from Atlantis clones to TF workspace managed
+
+This is not to suggest the revival of these PRs in their current state but to act as a reference for additional focused solutions. 
+
+## Solution: Clone once + TF_DATA_DIR
+
+Take PR #2921 and re-implement locks on the terraform DATA_DIR only and the entire base repo + PR for git clones. This will focus solely on git operations and Terraform Data Directories. 
+
+### Clone once
+
+We should be avoiding altogether re-cloning the repo unless absolutely necessary. We currently attempt to clone 3 times in a single command execution.
+
+1) Pre-Workflow Hooks
+2) <Import/Plan/Apply>CommandRunner
+	a) determintes projects to run plans on
+3) ProjectCommandRunner
+	a) clones before actual plan execution
+
+We should clone one initially for the entire repo/PR. This will empower multiple workspaces/plans alongside TF_DATA_DIR without needing to re-clone.
+
+Here are the cases that will trigger a clone:
+
+	1) Initial clone of repo
+ 	2) Force Reclone 
+  		a) Error when attempting fetch + pull to update existing local git repo
+    		b) File operation on a path that doesn't exist (data loss/accidental deletes)
+
+In all other situations, we should be utilizing Git for its delta compressions and performing less intensive fetch + pull operations.
+
+### TF_DATA_DIR
+
+Terraform currently contains its working directory data at `TF_DATA_DIR` which by default is `.terraform`. Utilizing this environmental override, we can store information about the individual project plans and backend state in separate data directories. This would allow for parallel plans not only per project, but across workspaces, in a single PR.
+
+The proposed new structure would pass through `.terraform-$workspace`. Workspace in this case refers to the terraform workspace specified on a project/repo. If one is not provided, it is set to `default`.
+
+### Locking
+
+There has also been a history of issues with colliding locks that disrupts parallel commands running. In this case, locking happens to avoid collison on file operations. We should be locking during Git operations but might not need to during workspace/plan file creation.
```

Comment:
The lock for git operations should cover the "lifetime" of the git operation and the tests performed to see if they are necessary:
 - lock
 - check for conditions ( "does directory exist", "are we are the right commit", "are we in sync", ..)
 - fix condition (make dir, check out, fetch, checkout, merge, ... )
 - unlock